src/toast/io/observation_hdf_load.py

```diff
@@ -0,0 +1,359 @@
+        # Only the root process reads
+        global_times = None
+        if obs.comm.group_rank == 0:
+            global_times = np.transpose(ds[:])
+
+        obs.intervals.create(field, global_times, times, fromrank=0)
+        del ds
+
+
+# FIXME:  Add options here to prune detectors on load.
+
+
+@function_timer
+def load_hdf5(
+    path,
+    comm,
+    process_rows=None,
+    meta=None,
+    detdata=None,
+    shared=None,
+    intervals=None,
+):
+    """Load an HDF5 observation.
+
+    By default, all detdata, shared, intervals, and noise models are loaded into
+    memory.  A subset of objects may be specified with a list of names
+    passed to the corresponding function arguments.
+
+    Args:
+        path (str):  The path to the file on disk.
+        comm (toast.Comm):  The toast communicator to use.
+        process_rows (int):  (Optional) The size of the rectangular process grid
+            in the detector direction.  This number must evenly divide into the size of
+            comm.  If not specified, defaults to the size of the communicator.
+        meta (list):  Only save this list of metadata objects.
+        detdata (list):  Only save this list of detdata objects.
+        shared (list):  Only save this list of shared objects.
+        intervals (list):  Only save this list of intervals objects.
+
+    Returns:
+        (Observation):  The constructed observation.
+
+    """
+    log = Logger.get()
+    env = Environment.get()
+    parallel = have_hdf5_parallel()
+
+    # Open the file and get the root group.  In both serial and parallel HDF5,
+    # multiple readers are supported.  We open the file with all processes to
+    # enable reading detector and shared data in parallel.
+    hf = None
+    hfgroup = None
+    file_version = None
+    if parallel:
+        hf = h5py.File(path, "r", driver="mpio", comm=comm.comm_group)
+        hgroup = hf
+
+    else:
+        hf = h5py.File(path, "r")
+        hgroup = hf
+
+    # Data format version check
+    file_version = hgroup.attrs["toast_format_version"]
+    if file_version != 0:
+        msg = f"HDF5 file '{path}' using unsupported data format {file_version}"
+        log.error(msg)
+        raise RuntimeError(msg)
+
+    # Observation properties
+    obs_name = hgroup.attrs["observation_name"]
+    obs_uid = hgroup.attrs["observation_uid"]
+    obs_dets = json.loads(hgroup.attrs["observation_detectors"])
+    obs_det_sets = None
+    if hgroup.attrs["observation_detector_sets"] != "NONE":
+        obs_det_sets = json.loads(hgroup.attrs["observation_detector_sets"])
+    obs_samples = hgroup.attrs["observation_samples"]
+    obs_sample_sets = None
+    if hgroup.attrs["observation_sample_sets"] != "NONE":
+        obs_sample_sets = [
+            [int(x) for x in y]
+            for y in json.loads(hgroup.attrs["observation_sample_sets"])
+        ]
+
+    # Instrument properties
+
+    # FIXME:  We should add save / load methods to these classes to
```

Comment:
I agree, this looks clunky but will do for now.

Reply:
Yes, there are definitely some refinements needed here.  Just trying to get something out the door for immediate use.